src/twisted/application/test/test_internet.py

```diff
@@ -1031,3 +1031,78 @@ def test_stopServiceOnStoppedService(self):
 
         self.assertIsNone(self.successResultOf(firstStopDeferred))
         self.assertIsNone(self.successResultOf(secondStopDeferred))
+
+
+    def test_onNewConnectionCalledWhenServiceStarts(self):
+        """
+        The C{onNewConnection} callable is called when the
+        L{ClientService.startService} is called.
+        """
+        newConnections = [0]
+        def onNewConnection(_proto):
+            newConnections[0] += 1
```

Comment:
It would also help to have a test which checks that whatever onNewConnection returns, in the end, its result is ignored and you get the original protocol

```
newConnectionCalls = []
def onNewConnection(protocol):
    newConnectionCalls.append(protocol)
    return object()

then in the test

self.assertEqual(1, len(newConnectionCalls))
self.assertIs(newConnectionCalls[0], cq.constructedProtocols[0])
```